LFS-471: Add support for nested subjects in the CSV import

nela-jankechova opened the pull request:
I did not complete this branch in time, this is a work in progress. The current problem is that `findSubjectId` method is not giving expected results.
[Jira link.](https://phenotips.atlassian.net/browse/LFS-471)

nela-jankechova committed:
Implement Starter Code

nela-jankechova committed:
LFS-471: Update Import

veronikaslc commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -610,28 +624,79 @@ private Resource findForm(final Node subject)
+    // If there are more entries in the subject types list, recurse with the new parent and new subject type.
+    // When the whole list of subject types is processed, return current subject as the subject to use for the row.
+    {
+        System.out.println("getOrCreateSubject");
```
If you do not need debugging lines anymore better remove them all for PR (same for commented-out code).

Commit:
LFS-471: Fixed findSubjectId; Removed debugging code

sashaandjic commented:
@veronikaslc how can I test this?

veronikaslc commented:
@sashaandjic not ready for test

Commit:
LFS-471: Fixed Subject hierarchy after import

veronikaslc commented:
To test download the test file from [LFS-471 JIRA issue](https://phenotips.atlassian.net/browse/LFS-471), build and start the app  and run this curl command from the place of where the file is:
`curl -u admin:admin http://localhost:8080/Forms -F ':questionnaire=/Questionnaires/Tumors' -F ':subjectType=/SubjectTypes/Patient' -F ':subjectType=/SubjectTypes/Tumor' -F ':data=@tumors.csv'`

IntegralProgrammer commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -607,31 +614,88 @@ private Resource findForm(final Node subject)
         final Iterator<Resource> results = this.resolver.get().findResources(query, "JCR-SQL2");
+
         if (results.hasNext()) {
-            return results.next();
+            cache.put(subjectKey, results.next().adaptTo(Node.class));
+            return results.next().adaptTo(Node.class);
         }
```
This seems like a problem - you are not returning what you are passing to `cache.put`. Are you sure this is what you want to do? Furthermore, if `results` only has a length of `1`, the line `return results.next().adaptTo(Node.class)` will end up calling `.adaptTo` on a null object and will generate a NPE.

IntegralProgrammer commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -152,9 +156,9 @@ protected void doPost(SlingHttpServletRequest request, SlingHttpServletResponse
-            this.subjectType.set(resourceResolver.getResource(
-                StringUtils.defaultIfBlank(request.getParameter(":subjectType"), "/SubjectTypes/Patient"))
-                .adaptTo(Node.class));
+            // this.subjectType.set(resourceResolver.getResource(
+            //     StringUtils.defaultIfBlank(request.getParameter(":subjectType"), "/SubjectTypes/Patient"))
+            //     .adaptTo(Node.class));
```
If this code is no longer used, it should be removed.

Commit:
LFS-471: Fixed Subject hierarchy after import
Fixed iterator issue and aded default subject type

sashaandjic commented:
tested

marta- reviewed (approved):
Looks fine, didn't test

OrdiNeu committed:
LFS-471: Add support for nested subjects in the CSV import

Refactor getOrCreateSubject into smaller functions to remove complexity
warnings

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -607,31 +616,138 @@ private Resource findForm(final Node subject)
+    // When the whole list of subject types is processed, return current subject as the subject to use for the row.
+    {
+        Node current = null;
+        String[] subjectTypes = request.getParameterValues(":subjectType");
```
This should be done only once at the start, the subject types should be put in a threadlocal, and then the request will not need to be passed to this method.

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -77,6 +81,7 @@
 )
 @SlingServletResourceTypes(resourceTypes = { "lfs/FormsHomepage" }, methods = { "POST" })
 @SlingServletName(servletName = "Data Import Servlet")
+@SuppressWarnings({"ClassFanOutComplexity"})
```
Not good. Can some functionality be moved in another class? Like, finding (or creating) a subject given an ID, a type, and a parent.

OrdiNeu replied:
I've reduced the class fan out complexity by hiding a few classes, but that might not've been the best way to deal with the warning.

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -607,31 +616,138 @@ private Resource findForm(final Node subject)
+        for (String type: subjectTypes) {
+            current = getOrCreateSubject(row, type, current);
+            if (current == null) {
+                return null;
```
Given that somatic variants can apply either to a region or a whole tumor, I think we should support this usecase: 3 types passed in the request (patient, tumor, tumor region), and rows may have 2 or 3 of the identifiers filled in. If we find 3, then the region is the target subject. If we find 2, and a blank for the region, then the tumor is the target subject.

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -607,31 +616,138 @@ private Resource findForm(final Node subject)
+    private Node findSubject(String subjectKey, String subjectId, Node typeNode, Node parent)
+    {
+        // Load a cached version if we already have one
+        Map<String, Node> cache = this.subjectTypeCache.get();
```
Huh, is `subjectTypeCache` used both for subjects and subject types? It should be renamed.

OrdiNeu replied:
Looking through the code, it doesn't seem to be used for subject types at all. I've renamed it in `71af351` :heavy_check_mark:

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -607,31 +616,138 @@ private Resource findForm(final Node subject)
+        }
+
+        Session session = this.resolver.get().adaptTo(Session.class);
+        Workspace workspace = session.getWorkspace();
```
I feel like this should be done only once per jmport, not for each subject search.

sashaandjic commented:
re-tested

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -640,17 +756,43 @@ private Resource getOrCreateSubject(final CSVRecord row)
+    {
+        String finalType = type.split("/SubjectTypes/")[1];
+        String query =
+            String.format("select n from [lfs:SubjectType] as n where n.label = '%s'",
```
Why? The subject type is supposed to be a valid path that we should just get from JCR with `this.resolver.get().getResource(type)`.

OrdiNeu replied:
I've switched this in `b09aaee` :heavy_check_mark:

OrdiNeu committed:
LFS-471: Add support for nested subjects in the CSV import

Apply code cleanup suggestions by @sdumitriu

OrdiNeu committed:
LFS-471: Add support for nested subjects in the CSV import

Resolve the subjectType directly instead of finding it via label

OrdiNeu committed:
LFS-471: Add support for nested subjects in the CSV import

Rename subjectTypeCache to subjectCache since it never actually stored
subjectTypes.

OrdiNeu committed:
LFS-471: Add support for nested subjects in the CSV import
Use the last filled out subject if blanks occur

OrdiNeu committed:
LFS-471: Add support for nested subjects in the CSV import
Fix up a compilation error

OrdiNeu commented:
@sdumitriu All of those changes have been addressed. Thanks!

sashaandjic commented:
re-tested

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/DataImportServlet.java:
```diff
@@ -607,31 +629,134 @@ private Resource findForm(final Node subject)
      *         {@code null} if a Subject identifier is not present in the row
      */
-    private Resource getOrCreateSubject(final CSVRecord row)
+    private Node getOrCreateSubject(final CSVRecord row, final SlingHttpServletRequest request)
```
I think `request` isn't needed anymore and could be removed.

sdumitriu committed:
[misc] Remove unused Request parameter from private methods

sdumitriu committed:
[misc] Codestyle improvements
